### PR TITLE
Initial code

### PR DESCRIPTION
**Pull request by lukicdarkoo**
This PR introduces basic functionalities required to create Webots animation and publish it to GitHub pages:
- it supports multiple worlds,
- it creates an animation by appending `animation_recorder` supervisor to worlds,
- it creates a simple single-page application with all words,
- it supports configuration through `webots.yaml`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Webots Animation Action
+
+This GitHub action creates a Webots animation of a simulation and publishes it to GitHub pages.
+
+<p align="center">
+  <img src="./assets/cover.png">
+</p>
+
+
+After each commit, Webots simulation will be recorded and published to `<username>.github.io/<repository>` as X3D animation.
```
Comment (omichel) [inline README.md]:
```suggestion
After each commit, Webots simulation will be recorded and published to `<username>.github.io/<repository>` as an X3D animation.
```

File: controllers/animation_recorder/animation_recorder.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    parser.add_argument('--duration', type=float, default=10, help='Duration of the animation in seconds')
+    parser.add_argument('--output', default='../../animation/index.html', help='Path at which the animation will be saved')
+    args = parser.parse_args()
+
+    robot = Supervisor()
+    timestep = int(robot.getBasicTimeStep())
+    robot.animationStartRecording(args.output)
+
+    step_i = 0
+    n_steps = (1e3 * args.duration) / robot.getBasicTimeStep()
```
Comment (omichel) [inline controllers/animation_recorder/animation_recorder.py]: 1000 is more readable than 1e3 for neophytes.
```suggestion
    n_steps = (1000 * args.duration) / robot.getBasicTimeStep()
```

File: scripts/template.html [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+</head>
+
+<body>
+    <div id="view3d"></div>
+    <div id="sidebar">
+        <h2>List of simulations</h2>
+    </div>
+</body>
+
+</html>
```
Comment (omichel) [inline scripts/template.html]: Missing final line feed.
```suggestion
</html>

```

File: scripts/template.html [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+        .simulation-item h3,
+        .simulation-item p {
+            margin: 0;
+        }
+    </style>
+    <script src="https://www.cyberbotics.com/jquery/1.11.3/jquery.min.js"></script>
+    <script src="https://www.cyberbotics.com/jquery-ui/1.11.4/jquery-ui.min.js"></script>
+    <script src="https://www.cyberbotics.com/jquery-dialogextend/2.0.4/jquery.dialogextend.min.js"></script>
+    <script src="https://www.cyberbotics.com/ace/1.2.0/ace.js"></script>
```
Comment (omichel) [inline scripts/template.html]: I believe this one is not useful.
```suggestion
```
Other JS dependencies like jquery-dialogextend may be useless as well (to be tested).

File: scripts/template.html [hunk 5]
```diff
@@ -0,0 +1,138 @@
+<!DOCTYPE html>
```
Comment (omichel) [inline scripts/template.html:1]: According to our coding style, we use 2 spaces for indentation in HTML, JavaScript and CSS:
https://github.com/cyberbotics/webots/wiki/Javascript-Coding-Style#cs401-use-2-spaces-indentation

File: scripts/run.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+            with open(world_file, 'r') as f:
+                world_content = f.read()
+
+            with open(world_file, 'w') as f:
+                f.write(world_content + animation_recorder_vrml)
+
+            command = Command(
+                f"xvfb-run webots --stdout --stderr --batch --mode=fast {world_file}")
+            command.run()
```
Comment (omichel) [inline scripts/run.py]: The use of the Command class seems overkill here, why not using the simpler subprocess?
```python
args = ['xvfb-run', 'webots', '--stdout', '--stderr', '--batch', '--mode=fast', world_file]
subprocess.call(args) 
```

File: entrypoint.sh [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+git config --global user.name "${NAME}"
+git config --global user.email "${ID}+${GITHUB_ACTOR}@users.noreply.github.com"
+git reset --hard
+git fetch
+git checkout gh-pages || git checkout -b gh-pages
+rm -rf $(ls -aI '.git') 2> /dev/null
+cp -r /tmp/animation/* .
+git add -A
+git commit -m "Update simulation"
```
Comment (omichel) [inline entrypoint.sh]: The message here doesn't seem relevant. What about this?
```suggestion
git commit -m "Updated animation"
```

File: scripts/run.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    for world in animation_config['worlds']:
+        for world_file in glob(world['file']):
+            description = ''
+            title = ''
+            with open(world_file, 'r') as f:
+                world_content = f.read()
+
+                # Parse `title`
+                title_expr = re.compile(
+                    r'title\s\"(.*?)\"', re.MULTILINE | re.DOTALL)
```
Comment (omichel) [inline scripts/run.py]: Why not fitting this on a single line?
```suggestion
                title_expr = re.compile(r'title\s\"(.*?)\"', re.MULTILINE | re.DOTALL)
```

File: scripts/run.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env python
```
Comment (omichel) [inline scripts/run.py]: This file requires Python >= 3.6 (at least for f strings). The shebang should reflect this and refer to at leas python3.

Coding style: Generally, we should not split an instruction across several lines unless we reach the line length limit.

File: scripts/template.html [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    <script>
+        const WORLD_LIST = { WORLD_LIST_PLACEHOLDER };
+        /*
+        const WORLD_LIST = [  
+            { description: 'Item description', title: 'Item title', name: 'world_name' },
+            { description: 'Item description', title: 'Item title', name: 'world_name' }
+        ];
+        */
+        var view = null;
+        var currentWorld = null;
```
Comment (omichel) [inline scripts/template.html]: We usually prefer the `let` keyword to the `var` one because of its better scope properties.

**Review by omichel — CHANGES_REQUESTED** [review]
It looks nice, although I have quite a few minor suggestions.

File: scripts/run.py (hunk 6 above)
Comment (lukicdarkoo) [reply scripts/run.py]: I saw `Command` has some nice features, so I wanted to use it e.g. parallel execution. However, I guess it is a better idea to keep it simple for now.

Comment (omichel) [reply scripts/run.py]: Yes, for now it looks overkill. And also I don't like too much copying/pasting a class this way for maintenance reasons.

File: scripts/run.py (hunk 8 above)
Comment (lukicdarkoo) [reply scripts/run.py]: The editor applied default line length :( Reconfigured

File: scripts/template.html (hunk 4 above)
Comment (lukicdarkoo) [reply scripts/template.html]: Doesn't work without ace.js. Deleting any of JS dependencies causes webots.js to fail.

Comment (omichel) [reply scripts/template.html]: Okay, so let's keep them for now (although we should find a way some day to get rid of them as they are not really used in this case).

File: README.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
+# Webots Animation Action
+
+This GitHub action creates a Webots animation of a simulation and publishes it to GitHub pages.
+
+<p align="center">
+  <img src="./assets/cover.png">
```
Comment (omichel) [inline README.md:6]: It is a pity we don't see the animation slider on this picture. Could you add it?

**Review by omichel — CHANGES_REQUESTED** [review]
Thank you.
I just have one last comment.

File: README.md (hunk 11 above)
Comment (lukicdarkoo) [reply README.md:6]: I think it is more descriptive now. However, adding a slider would be too many details?

Comment (omichel) [reply README.md:6]: Ok, let's go ahead with this.

**Review by omichel — APPROVED** [review]
Thank you.

File: Dockerfile [hunk 12]
```diff
@@ -0,0 +1,10 @@
+FROM cyberbotics/webots
```
Comment (DavidMansolino) [inline Dockerfile]: It's probably safer to specify a version here (so we are sure which one we are using).

File: .gitignore [hunk 13]
```diff
@@ -0,0 +1 @@
+scripts/__pycache__
```
Comment (DavidMansolino) [inline .gitignore]: This gitignore can be moved in the `scripts` folder.

File: README.md [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+on: [push]
+
+jobs:
+  record:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out the repo
+        uses: actions/checkout@v2
+      - name: Record and deploy the animation
+        uses: cyberbotics/webots-animation-action@master
```
Comment (DavidMansolino) [inline README.md:31]: Can you really reference a branch?
From my previous test I was only able to publish the action by creating a tag and an associated release (which we probably don't want to call `master` :wink:)
We should probably use version names such as the ones from the official actions (e.g. https://github.com/marketplace/actions/checkout or https://github.com/marketplace/actions/setup-python).

File: Dockerfile [hunk 15]
```diff
@@ -0,0 +1,10 @@
+FROM cyberbotics/webots 	
+
+RUN apt-get update && \ 	
+    apt-get install -y git rsync python3-yaml python3-jinja2 jq
```
Comment (DavidMansolino) [inline Dockerfile]: It seems you are not using jinja2 but raw string replace no?

Comment (DavidMansolino) [reply Dockerfile]: Same for `rsync` actually.

File: README.md [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+After each commit, Webots simulation will be recorded and published to `<username>.github.io/<repository>` as an X3D animation.
+In your browser, you can move around and zoom while the animation is playing.
+
+## Workflow
+
+Here is a simple GitHub workflow snippet which utilizes the action:
+```yaml
+name: Record animation
+
+on: [push]
```
Comment (DavidMansolino) [inline README.md]:
```suggestion
on:
  push:
    branches: 
      - master
```
This is just a suggestion, but the user may want that the animation(s) represent the state of master.

**Review by DavidMansolino — COMMENTED** [review]
I really like your idea!!
Just a few suggestions/questions.

File: README.md (hunk 14 above)
Comment (lukicdarkoo) [reply README.md:31]: It works, see here :)
https://github.com/cyberbotics/webots-animation-template/tree/develop
I think you can put anything that can be associated with a commit, like hash, branch, or tag.

Once we are satisfied with API and implementation we can add a tag.

File: Dockerfile (hunk 15 above)
Comment (lukicdarkoo) [reply Dockerfile]: > It seems you are not using jinja2 but raw string replace no?

True, I started with jinja2, but there was no need for it.

File: README.md (hunk 16 above)
Comment (lukicdarkoo) [reply README.md]: Good idea!

File: README.md (hunk 14 above)
Comment (DavidMansolino) [reply README.md:31]: Good to know, that's indeed a wonderful feature for the development! Then let's keep it like this for now and create the official 'release' to be integrated in the marketplace once it is fully ready.

**Pull request merged**